source/extensions/filters/http/common/compressor/compressor.h

```diff
@@ -0,0 +1,172 @@
+  COUNTER(no_accept_header)                                                                        \
+  COUNTER(header_identity)                                                                         \
+  COUNTER(header_gzip)                                                                             \
+  COUNTER(header_compressor_used)                                                                  \
+  COUNTER(header_compressor_overshadowed)                                                          \
+  COUNTER(header_wildcard)                                                                         \
+  COUNTER(header_not_valid)                                                                        \
+  COUNTER(total_uncompressed_bytes)                                                                \
+  COUNTER(total_compressed_bytes)                                                                  \
+  COUNTER(content_length_too_small)                                                                \
+  COUNTER(not_compressed_etag)
+
+/**
+ * Struct definition for compressor stats. @see stats_macros.h
+ */
+struct CompressorStats {
+  ALL_COMPRESSOR_STATS(GENERATE_COUNTER_STRUCT)
+};
+
+class CompressorFilterConfig {
+public:
+  CompressorFilterConfig() = delete;
+  virtual ~CompressorFilterConfig() = default;
+
+  virtual const std::string featureName() const PURE;
+  virtual std::unique_ptr<Compressor::Compressor> makeCompressor() PURE;
+
+  Runtime::Loader& runtime() { return runtime_; }
+  CompressorStats& stats() { return stats_; }
+  const StringUtil::CaseUnorderedSet& contentTypeValues() const { return content_type_values_; }
+  bool disableOnEtagHeader() const { return disable_on_etag_header_; }
+  bool removeAcceptEncodingHeader() const { return remove_accept_encoding_header_; }
+  uint32_t minimumLength() const { return content_length_; }
+  const std::string contentEncoding() const { return content_encoding_; };
+  const std::map<std::string, uint32_t> registeredCompressors() const;
+
+protected:
+  CompressorFilterConfig(
+      const envoy::extensions::filters::http::compressor::v3::Compressor& compressor,
+      const std::string& stats_prefix, Stats::Scope& scope, Runtime::Loader& runtime,
+      const std::string& content_encoding);
+
+private:
+  static StringUtil::CaseUnorderedSet
+  contentTypeSet(const Protobuf::RepeatedPtrField<std::string>& types);
+
+  static uint32_t contentLengthUint(Protobuf::uint32 length);
+
+  static CompressorStats generateStats(const std::string& prefix, Stats::Scope& scope) {
+    return CompressorStats{ALL_COMPRESSOR_STATS(POOL_COUNTER_PREFIX(scope, prefix))};
+  }
+
+  uint32_t content_length_;
+
+  StringUtil::CaseUnorderedSet content_type_values_;
+  bool disable_on_etag_header_;
+  bool remove_accept_encoding_header_;
+  CompressorStats stats_;
+  Runtime::Loader& runtime_;
+  std::string content_encoding_;
```

Comment:
nit: please constify all possible member variables here and below, using the initializer list where possible if not already doing so.